File: cpp/leetcode/302.cpp
class Solution {
public:
    int minArea(vector<vector<char>>& image, int x, int y) {
        if(image.empty() || image[0].empty()) return 0;
        int m = image.size(), n = image[0].size();
        int top = row(0, x, 0, n, true, image);
        int bot = row(x+1, m, 0, n, false, image);
        int left = col(0, y, top, bot, true, image);
        int right = col(y + 1, n, top, bot, false, image);
        return (right - left) * (bot - top);
    }
    
    int row(int top, int bot, int left, int right, bool flag, vector<vector<char>> &image){
        while(top != bot){
            int i = left, mid = (bot - top)/2 + top;
            while(i < right && image[mid][i] == '0') i++;
            if(i < right == flag) bot = mid;
            else top = mid + 1;
        }
        return top;
    }
    
    int col(int top, int bot, int left, int right, bool flag, vector<vector<char>> &image){
        while(top != bot){
            int i = left, mid = (bot - top)/2 + top;
            while(i < right && image[i][mid] == '0') i++;
            if(i < right == flag) bot = mid;
            else top = mid + 1;
        }
        return top;
    }
};

File: python/leetcode/302.py
class Soluiton:
    def minArea(self, image, x, y):
        m, n = len(image), len(image[0])
        top = self.searchRows(image, 0, x, True)
        bottom = self.searchRows(image, x + 1, m, False)
        left = self.searchColumns(image, 0, y, top, bottom, True)
        right = self.searchColumns(image, y + 1, n, top, bottom, False)
        return (right - left) * (bottom - top)

    def searchRows(self, image, i, j, opt):
        while i != j:
            m = (i + j) / 2
            if ('1' in image[m]) == opt: j = m  # j = m not m - 1
            else: i = m + 1
        return i

    def searchColumns(self, image, i, j, top, bot, opt):
        while i != j:
            m = (i + j) / 2
            if any(image[k][m]=='1' for k in xrange(top, bot))==opt: j = m
            else: i = m + 1
        return i